src/api/cvc4cpp.cpp

```diff
@@ -4230,6 +4436,115 @@ Term Solver::ensureTermSort(const Term& term, const Sort& sort) const
+
+  Expr res = d_exprMgr->mkBoundVar(symbol, *sort.d_type);
+  (void)res.getType(true); /* kick off type checking */
+
+  d_smtEngine->declareSygusVar(symbol, res, *sort.d_type);
+
+  return res;
+
+  CVC4_API_SOLVER_TRY_CATCH_END;
+}
+
+Grammar Solver::mkGrammar(const std::vector<Term>& boundVars,
+                          const std::vector<Term>& ntSymbols) const
+{
+  CVC4_API_ARG_SIZE_CHECK_EXPECTED(!ntSymbols.empty(), ntSymbols)
+      << "non-empty vector";
+
+  return Grammar(this, boundVars, ntSymbols);
+}
+
+Term Solver::synthFun(const std::string& symbol,
+                      const std::vector<Term>& boundVars,
+                      Sort sort) const
+{
+  return synthFunInternal(symbol, boundVars, sort);
+}
+
+Term Solver::synthFun(const std::string& symbol,
+                      const std::vector<Term>& boundVars,
+                      Sort sort,
+                      Grammar g) const
+{
+  return synthFunInternal(symbol, boundVars, sort, false, &g);
+}
+
+Term Solver::synthInv(const std::string& symbol,
+                      const std::vector<Term>& boundVars) const
+{
+  return synthFunInternal(symbol, boundVars, d_exprMgr->booleanType(), true);
+}
+
+Term Solver::synthInv(const std::string& symbol,
+                      const std::vector<Term>& boundVars,
+                      Grammar g) const
+{
+  return synthFunInternal(
+      symbol, boundVars, d_exprMgr->booleanType(), true, &g);
+}
+
+Term Solver::synthFunInternal(const std::string& symbol,
+                              const std::vector<Term>& boundVars,
+                              const Sort& sort,
+                              bool isInv,
+                              Grammar* g) const
+{
+  CVC4_API_SOLVER_TRY_CATCH_BEGIN;
+  CVC4_API_ARG_CHECK_EXPECTED(!symbol.empty(), sort) << "non-empty string";
+  CVC4_API_ARG_CHECK_EXPECTED(!sort.isNull(), sort) << "non-null sort";
+
+  std::vector<Type> var_types;
+  for (const Term& p : boundVars)
+  {
+    var_types.push_back(p.d_expr->getType());
+  }
+
+  Type fun_type = var_types.empty()
+                      ? *sort.d_type
+                      : d_exprMgr->mkFunctionType(var_types, *sort.d_type);
+
+  Expr fun = d_exprMgr->mkBoundVar(symbol, fun_type);
+  (void)fun.getType(true); /* kick off type checking */
+
+  d_smtEngine->declareSynthFun(symbol,
+                               fun,
+                               g == nullptr ? fun_type : *g->ret().d_type,
+                               isInv,
+                               termVectorToExprs(boundVars));
+
+  return fun;
+
+  CVC4_API_SOLVER_TRY_CATCH_END;
+}
+
+void Solver::addConstraint(const Term& term) const
+{
+  d_smtEngine->assertSygusConstraint(*term.d_expr);
+}
+
+void Solver::addInvariantConstraint(const Term& inv,
+                                    const Term& pre,
+                                    const Term& trans,
+                                    const Term& post) const
+{
```

Comment:
Ditto, check for non-null `inv`, `pre`, `trans`, `post`.

Reply:
Done!